content/article/passing-by.md

```diff
@@ -0,0 +1,237 @@
+```perl
+my $str = "foo";
+
+# Get reference
+my $ref = \$str;
+
+print "$$ref\n";
+```
+Or 
+```perl
+my @array = ( "foo", "bar", "baz" );
+
+# Get reference
+my $ref = \@array;
+
+# Derefrence with ->
+print "$ref->[0]\n";
+```
+Both of them will print:
+```
+foo
+```
+
+You can even obtain a **reference of a reference** and later **dereference multiple times**. It enables the possibility to produce weird/dumb things like this:
+```perl
+my $str = "bazinga";
+
+# Get a ref on a ref on a ref on a...
+my $rrrrrrrrrref = \\\\\\\\\\$str;
+
+# De-de-de-...-de-dereference
+print "$$$$$$$$$$$rrrrrrrrrref\n";
+```
+
+But I repeat, don't do this.
+![](/images/passing-by/weird.png)
+
+## Modifications of variables into subs
+Inside subs, editing the variables won't usually ([\*](https://stackoverflow.com/a/5746000)) affect the arguments.
+```perl
+sub bycopy {
+    my $cp = shift;
+    $cp = "modified";
+}
+
+my $var = "not modified";
+bycopy($var);
+
+print "$var\n";
+```
+(for the curious, using `&` or a `prototype` won't change the game)
+
+That prints:
+```
+not modified
+```
+The modification is not visible outside the sub because in order to propagate the change, it should be returned to caller with a `return $cp;` and assigned like `$var = bycopy($var)` (actually even the return is optional because of "default variable magic").
+
+Obviously, this limitation is usually a good idea (side effect...), but here I was actually trying on purpose to modify the variable even for the caller...
+
+![](/images/passing-by/students.jpg)
+
+
+To be able to modify the variable inside the sub so that it will be propagated to caller, I can use a reference:
+```perl
+sub byref {
+    my $ref = shift;
+    $$ref = "modified";
+}
+
+my $variable = "not modified";
+byref(\$variable);
+
+print "$variable\n";
+```
+That prints:
+```
+modified
+```
+This way, I pass the variable itself, whatever it is (scalar, array, hash...), I can edit it and it will persist.
+
+But there is an easier and obvious  way that I do not talked until now, the default array `@_` is actually passed by reference!
+```perl
+sub bydefaultarray {
+    $_[0] = "modified";
+}
+
+my $variable = "not modified";
+bydefaultarray($variable);
+
+print "$variable\n";
+```
+And it is well modified:
+```
+modified
+```
+
+You only have to take care to work directly on `@_` (or aliases made with care) and do not use a function like `shift` that will give you *only* a copy.
+
+## Implicit conversion to reference with prototype
+Then there is another thing to know around subs and references: the usage of prototype.
+Prototypes change the way perl behaves and it allows for instance a sub to receive an array and to implicitely convert it to a reference.
+
+The following example forces the array to its reference instead of flattening it:
+```perl
+sub proto(\@) {
+    my $ref = shift; # It is the array ref, not the first item
+    $ref->[0] = "modified";
```

Comment:
I'd use unique values for each element to distinguish the case where one assignment affected all elements.

Reply:
OK